text/0000-delegation-of-implementation.md

```diff
@@ -0,0 +1,235 @@
+
+## Partial delegation 
+
+If we consider this code
+```rust
+// from rust/src/libsyntax/attr.rs
+impl AttrMetaMethods for Attribute {
+    fn check_name(&self, name: &str) -> bool {
+        let matches = name == &self.name()[..];
+        if matches {
+            mark_used(self);
+        }
+        matches
+    }
+    fn name(&self) -> InternedString { self.meta().name() }
+    fn value_str(&self) -> Option<InternedString> {
+        self.meta().value_str()
+    }
+    fn meta_item_list(&self) -> Option<&[P<MetaItem>]> {
+        self.node.value.meta_item_list()
+    }
+    fn span(&self) -> Span { self.meta().span }
+}
+```
+we can identify the recurring expression `self.meta()` but 2 of the 5 methods are more complex. This heterogeneity can be handled simply if we allow partial delegation like in
+```rust
+impl AttrMetaMethods for Attribute use self.meta() {
+    fn check_name(&self, name: &str) -> bool {
+        let matches = name == &self.name()[..];
+        if matches {
+            mark_used(self);
+        }
+        matches
+    }
+    fn meta_item_list(&self) -> Option<&[P<MetaItem>]> {
+        self.node.value.meta_item_list()
+    }
+}
+```
+Only missing methods are automatically implemented.
+
+In some other cases the compiler just cannot generate the appropriate method. For example when `self` is moved rather than passed by reference, unless the delegating expression produces a result that can itself be moved the borrow checker will complain. In that kind of situations the developer can again provide a custom implementation where necessary and let the compiler handle the rest of the methods.
+
+## Delegation for other parameters 
+
+If `Self` is used for other parameters, everything works nicely and no specific treatment is required.
+```rust
+// from rust/src/libcollections/btree/map.rs
+impl<K: PartialOrd, V: PartialOrd> PartialOrd for BTreeMap<K, V> {
+    fn partial_cmp(&self, other: &BTreeMap<K, V>) -> Option<Ordering> {
+        self.iter().partial_cmp(other.iter())
+    }
+}
+```
+becomes 
+```rust
+impl<K: PartialOrd, V: PartialOrd> PartialOrd for BTreeMap<K, V> use self.iter();
+```
+
+## Associated types/constants 
+
+Unless explicitly set associated types and constants should default to the surrogate implementation value of the corresponding items.
```

Comment:
How could a type explicitly set associated types different from the surrogate implementation without introducing a type error?